aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -46,71 +100,22 @@ void spmm_reduce_kernel_impl(
   utils::parallel_sparse_csr(csr_data, M, nnz, [&](int64_t begin, int64_t end) {
-    int64_t row_start, row_end, c;
+    int64_t row_start, row_end;
     for (const auto m : c10::irange(begin, end)) {
       row_start = csr_data[m];
       row_end = csr_data[m + 1];
 
       scalar_t* out_ptr = out_data + m * K;
 
-      constexpr int64_t kVecSize = Vec::size();
-      constexpr int64_t kVLEN = kVecSize * 4;
-      constexpr int64_t CHUNK_SIZE = 16;
-
       // step 1: reinit the output row for reduce type 'amax' and 'amin'
       int64_t count = row_end - row_start;
       if (count != 0) {
         init<scalar_t, reduce>(out_ptr, K, /*include_self*/false);
       }
 
       // step 2: reduce, do blocking on rowwise to reduce write memory bandwidth
-      for (int64_t e0 = row_start; e0 < row_end; e0 += CHUNK_SIZE) {
-        int64_t e1 = std::min(e0 + CHUNK_SIZE, row_end);
-
-        int64_t k = 0;
-        for (; k < K - (K % kVLEN); k += kVLEN) {
-          Vec out_vec0 = Vec::loadu(out_ptr + k);
-          Vec out_vec1 = Vec::loadu(out_ptr + k + kVecSize);
-          Vec out_vec2 = Vec::loadu(out_ptr + k + kVecSize * 2);
-          Vec out_vec3 = Vec::loadu(out_ptr + k + kVecSize * 3);
-          for (const auto e : c10::irange(e0, e1)) {
-            c = col_data[e];
-            scalar_t val = val_data[e];
-            scalar_t* other_ptr = other_data + c * K + k;
-
-            out_vec0 = update<Vec, reduce>(out_vec0, Vec::loadu(other_ptr) * Vec(val));
-            out_vec1 = update<Vec, reduce>(out_vec1, Vec::loadu(other_ptr + kVecSize) * Vec(val));
-            out_vec2 = update<Vec, reduce>(out_vec2, Vec::loadu(other_ptr + kVecSize * 2) * Vec(val));
-            out_vec3 = update<Vec, reduce>(out_vec3, Vec::loadu(other_ptr + kVecSize * 3) * Vec(val));
-          }
-          out_vec0.store(out_ptr + k);
-          out_vec1.store(out_ptr + k + kVecSize);
-          out_vec2.store(out_ptr + k + kVecSize * 2);
-          out_vec3.store(out_ptr + k + kVecSize * 3);
-        }
-        for (; k < K - (K % kVecSize); k += kVecSize) {
-          Vec out_vec = Vec::loadu(out_ptr + k);
-          for (const auto e : c10::irange(e0, e1)) {
-            c = col_data[e];
-            scalar_t val = val_data[e];
-            scalar_t* other_ptr = other_data + c * K;
-            out_vec = update<Vec, reduce>(out_vec, Vec::loadu(other_ptr + k) * Vec(val));
-          }
-          out_vec.store(out_ptr + k);
-        }
-        for (; k < K; k++) {
-          scalar_t out_val = out_ptr[k];
-          for (const auto e : c10::irange(e0, e1)) {
-            c = col_data[e];
-            scalar_t val = val_data[e];
-            scalar_t* other_ptr = other_data + c * K;
-            out_val = update<scalar_t, reduce>(out_val, other_ptr[k] * val);
-          }
-          out_ptr[k] = out_val;
-        }
-      }
+      _update<scalar_t, index_t, reduce>(out_ptr, row_start, row_end, col_data, val_data, other_data, K);
```

Comment:
the same story as scatter_reduce here. we need to allocate a thread local temp buffer of fp32 to replace the `out_ptr` here. and do accumulation on the temp buffer and then write to output (also convert fp32 to bf16).

you can use `constexpr is_reduced_floating_point = ...` to bypass some steps on fp32.